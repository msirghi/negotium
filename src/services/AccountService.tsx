import Requests from '../common/requests/request';
import { BASE_API_URL_V1 } from '../common/constants/constants';
import { HttpMethod } from '../common/requests/types';
import ServiceResultFactory from '../common/requests/serviceResultFactory';

const getUserMetadata = () => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/metadata`,
    HttpMethod.GET
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const updateUserTheme = (theme: string) => {
  return Requests.restApiCallWithBearer(
    `${BASE_API_URL_V1}/metadata`,
    HttpMethod.PATCH,
    { theme }
  )
    .then(ServiceResultFactory.fromResponse)
    .catch(ServiceResultFactory.fromError);
};

const AccountService = {
  getUserMetadata,
  updateUserTheme,
};

export default AccountService;
