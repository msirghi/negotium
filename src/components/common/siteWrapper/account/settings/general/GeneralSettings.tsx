import { makeStyles } from '@mui/styles';
import { Divider, MenuItem, TextField } from '@mui/material';
import {
  HOME_VIEW_LIST,
  SUPPORTED_LANGUAGES,
  TIME_FORMATS,
} from '../../../../../../common/constants/constants';
import { useTranslation } from 'next-i18next';
import { Box } from '@mui/system';
import { useCommonStyles } from '../../styles';
import { useIsMobile } from '../../../../../../common/hooks/common/useIsMobile';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '../../../../../../redux/store';
import AccountService from '../../../../../../services/AccountService';
import { setLanguage } from '../../../../../../redux/account/accountSlice';
import { useRouter } from 'next/router';

const useStyles = makeStyles({
  dropdown: {
    minWidth: '50%',
  },
});

export const GeneralSettings = () => {
  const commonClasses = useCommonStyles();
  const classes = useStyles();
  const { t } = useTranslation('settings');
  const isMobile = useIsMobile();
  const router = useRouter();
  const userMetadata = useSelector(
    (state: RootState) => state.account.metadata
  );
  const dispatch = useDispatch();

  const updateLanguage = async (evt: { target: { value: string } }) => {
    const language = evt.target.value;
    dispatch(setLanguage(language));
    await AccountService.updateUserLanguage(language);
    await router.push(router.route, router.route, { locale: language });
  };

  return (
    <div>
      <Box className={commonClasses.sectionTitle}>{t('language')}</Box>
      <Box className={commonClasses.sectionBody}>
        <TextField
          select
          value={userMetadata.language}
          size={'small'}
          className={classes.dropdown}
          fullWidth={isMobile}
          onChange={updateLanguage}
        >
          {SUPPORTED_LANGUAGES.map(({ code, title }) => {
            return (
              <MenuItem key={code} value={code}>
                {title}
              </MenuItem>
            );
          })}
        </TextField>
      </Box>

      <Box className={commonClasses.sectionTitle}>{t('homeView')}</Box>
      <Box className={commonClasses.sectionBody}>
        <TextField
          select
          value={'Inbox'}
          size={'small'}
          className={classes.dropdown}
          fullWidth={isMobile}
        >
          {HOME_VIEW_LIST.map(({ title }) => {
            return (
              <MenuItem key={title} value={title}>
                {title}
              </MenuItem>
            );
          })}
        </TextField>
      </Box>

      <Box className={commonClasses.sectionTitle}>
        <Divider />
      </Box>

      <Box className={commonClasses.sectionTitle}>{t('timeFormat')}</Box>
      <Box className={commonClasses.sectionBody}>
        <TextField
          select
          value={'24h'}
          size={'small'}
          className={classes.dropdown}
          fullWidth={isMobile}
        >
          {TIME_FORMATS.map(({ title, key }) => {
            return (
              <MenuItem key={key} value={key}>
                {title}
              </MenuItem>
            );
          })}
        </TextField>
      </Box>
    </div>
  );
};
