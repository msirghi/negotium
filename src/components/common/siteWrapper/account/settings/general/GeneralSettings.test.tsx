import renderer from 'react-test-renderer';
import { MockReduxProvider } from '../../../../../../common/tests/TestUtils';
import { GeneralSettings } from './GeneralSettings';
import { mount } from 'enzyme';
import { TextField } from '@mui/material';
import { act } from '@testing-library/react';
import AccountService from '../../../../../../services/AccountService';

jest.mock('next/router', () => ({
  useRouter: () => ({
    route: 'inbox',
    query: {},
  }),
}));

describe('GeneralSettings', () => {
  const renderComponent = () => {
    return (
      <MockReduxProvider
        reduxStore={{
          account: {
            metadata: {
              language: 'en',
            },
          },
        }}
      >
        <GeneralSettings />
      </MockReduxProvider>
    );
  };

  it('should match the snapshot', () => {
    const tree = renderer.create(renderComponent());
    expect(tree).toMatchSnapshot();
  });

  it('should call the service on language change', () => {
    AccountService.updateUserLanguage = jest.fn(() => Promise.resolve()) as any;
    const wrapper = mount(renderComponent());
    const languageSelector = wrapper.find(TextField).at(0) as any;

    act(() => {
      languageSelector.props().onChange({ target: { value: 'ru' } });
    });
    expect(AccountService.updateUserLanguage).toBeCalled();
  });
});
