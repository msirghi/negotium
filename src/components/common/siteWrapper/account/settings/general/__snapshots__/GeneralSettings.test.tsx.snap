// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeneralSettings should match the snapshot 1`] = `
<div>
  <div
    className="makeStyles-sectionTitle-2 MuiBox-root css-0"
  >
    language
  </div>
  <div
    className="makeStyles-sectionBody-3 MuiBox-root css-0"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-dropdown-4 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          English
        </div>
        <input
          aria-hidden={true}
          autoFocus={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="en"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1a1ihhj"
          >
            <span
              className="notranslate"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-sectionTitle-2 MuiBox-root css-0"
  >
    homeView
  </div>
  <div
    className="makeStyles-sectionBody-3 MuiBox-root css-0"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-dropdown-4 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          Inbox
        </div>
        <input
          aria-hidden={true}
          autoFocus={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="Inbox"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1a1ihhj"
          >
            <span
              className="notranslate"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    className="makeStyles-sectionTitle-2 MuiBox-root css-0"
  >
    <hr
      className="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
    />
  </div>
  <div
    className="makeStyles-sectionTitle-2 MuiBox-root css-0"
  >
    timeFormat
  </div>
  <div
    className="makeStyles-sectionBody-3 MuiBox-root css-0"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-dropdown-4 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
    >
      <div
        className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiSelect-root css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        onClick={[Function]}
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          className="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          role="button"
          tabIndex={0}
        >
          13:00
        </div>
        <input
          aria-hidden={true}
          autoFocus={false}
          className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
          disabled={false}
          onAnimationStart={[Function]}
          onChange={[Function]}
          required={false}
          tabIndex={-1}
          value="24h"
        />
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
          data-testid="ArrowDropDownIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
        <fieldset
          aria-hidden={true}
          className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            className="css-1a1ihhj"
          >
            <span
              className="notranslate"
              dangerouslySetInnerHTML={
                Object {
                  "__html": "&#8203;",
                }
              }
            />
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
